TypeScript_ESLint.snb.md

```diff
@@ -2,130 +2,188 @@
+- `messages`: An object keying ID strings to full messages that may be reported by the rule
+
+`no-extra-non-null-assertion`'s `meta` indicates it has an auto-fixer for code issues and that it may report _`Forbidden extra non-null assertion`_ if it detects an issue.
 
-One distinction between TypeScript ESLint and vanilla ESLint is that we have access to the TypeScript type checker and can therefore define rules that make use of type information.
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint@3eab889022c9d1617f275017d6951f663ea57f24/-/blob/packages/eslint-plugin/src/rules/no-extra-non-null-assertion.ts?L6-17
 
-The `for-in-array` rule uses the type checker to determine if we are dealing with an array type:
+### Rule Creator Functions
 
-https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts?L23-24
+A rule's `create` function returns an object whose keys are AST queries and whose values are functions that will be run whenever a node matching the corresponding query is found.
+The AST query keys use the [ESQuery](https://github.com/estools/esquery) library, which allows using syntax similar to CSS selectors for querying types of nodes.
 
-https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts?L32-35
+`no-extra-non-null-assertion` creates three node queries in its returned object:
 
-## Dual ASTs
+1. Any `!` non-null assertion within another `!` non-null assertion: e.g. `scream!!.length`
+2. Any `!` non-null assertion just before an `?.` optional chained property: e.g. `scream!?.length`
+3. Any `!` non-null assertion just before an `?.` optional chained call: e.g. `scream!?.()`
 
-In addition to invoking the type checker, the `for-in-array` rule also makes use of **two ASTs**:
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint@3eab889022c9d1617f275017d6951f663ea57f24/-/blob/packages/eslint-plugin/src/rules/no-extra-non-null-assertion.ts?L19-39
 
-1. the ESTree representation used by ESLint
-2. the TypeScript AST emitted by the TypeScript compiler.
+Note that any AST node names prefixed with `TS`, including `TSNonNullExpression`, are TypeScript-specific nodes.
+These are available to ESLint rules when using [`@typescript-eslint/parser`](http://npmjs.com/package/@typescript-eslint/parser) instead of ESLint's built-in parser.
+The `@typescript-eslint/utils` package exports those types under the `TSESTree` namespace.
 
-The `node` value passed as an argument to the rule function is from AST #1, an instance of [`TSESTree.Node`](https://sourcegraph.com/npm/typescript-eslint/types/-/blob/dist/generated/ast-spec.d.ts?L641):
+The corresponding TypeScript types for all the possible node types are stored in a [`@typescript-eslint/types`](https://npmjs.com/package/@typescript-eslint/types) package.
```

Comment:
@bradzacher the `types` package's README.md is pretty sparse. Should we even mention that package at all? Is there a better way to refer to the types?